uspace/lib/drv/generic/remote_usbhc.c

```diff
@@ -184,184 +220,283 @@ errno_t usbhc_write(async_exch_t *exch, usb_address_t address,
+		async_answer_0(data_callid, EINVAL);
+		return EINVAL;
 	}
 
-	const errno_t rc = hc_iface->read(
-	    fun, target, setup, trans->buffer, size, callback_in, trans);
+	if ((err = async_share_out_finalize(data_callid, &trans->request.buffer.virt)))
+		return err;
 
-	if (rc != EOK) {
-		async_answer_0(trans->data_caller, rc);
-		async_answer_0(callid, rc);
-		async_transaction_destroy(trans);
+	/*
+	 * As we're going to get physical addresses of the mapping, we must make
+	 * sure the memory is actually mapped. We must do it right now, because
+	 * the area might be read-only or write-only, and we may be unsure
+	 * later.
+	 */
+	if (flags & AS_AREA_READ) {
+		char foo = 0;
+		volatile const char *buf = trans->request.buffer.virt + trans->request.offset;
+		for (size_t i = 0; i < size; i += PAGE_SIZE)
+			foo += buf[i];
+	} else {
+		volatile char *buf = trans->request.buffer.virt + trans->request.offset;
+		for (size_t i = 0; i < size; i += PAGE_SIZE)
+			buf[i] = 0xff;
 	}
+
+	return EOK;
 }
 
-void remote_usbhc_write(
-    ddf_fun_t *fun, void *iface, ipc_callid_t callid, ipc_call_t *call)
+void remote_usbhc_transfer(ddf_fun_t *fun, void *iface, ipc_callid_t callid, ipc_call_t *call)
 {
 	assert(fun);
 	assert(iface);
 	assert(call);
 
-	const usbhc_iface_t *hc_iface = iface;
+	const usbhc_iface_t *usbhc_iface = iface;
 
-	if (!hc_iface->write) {
+	if (!usbhc_iface->transfer) {
 		async_answer_0(callid, ENOTSUP);
 		return;
 	}
 
-	const usb_target_t target = { .packed = DEV_IPC_GET_ARG1(*call) };
-	const size_t data_buffer_len = DEV_IPC_GET_ARG2(*call);
-	const uint64_t setup =
-	    ((uint64_t)DEV_IPC_GET_ARG3(*call)) |
-	    (((uint64_t)DEV_IPC_GET_ARG4(*call)) << 32);
-
 	async_transaction_t *trans = async_transaction_create(callid);
 	if (trans == NULL) {
 		async_answer_0(callid, ENOMEM);
 		return;
 	}
 
-	size_t size = 0;
-	if (data_buffer_len > 0) {
-		const errno_t rc = async_data_write_accept(&trans->buffer, false,
-		    1, USB_MAX_PAYLOAD_SIZE,
-		    0, &size);
+	errno_t err = EPARTY;
 
-		if (rc != EOK) {
-			async_answer_0(callid, rc);
-			async_transaction_destroy(trans);
-			return;
-		}
+	ipc_callid_t data_callid;
+	size_t len;
+	if (!async_data_write_receive(&data_callid, &len)
+	    || len != sizeof(trans->request)) {
+		async_answer_0(data_callid, EINVAL);
+		goto err;
 	}
 
-	const errno_t rc = hc_iface->write(
-	    fun, target, setup, trans->buffer, size, callback_out, trans);
+	if ((err = async_data_write_finalize(data_callid,
+			    &trans->request, sizeof(trans->request))))
+		goto err;
 
-	if (rc != EOK) {
-		async_answer_0(callid, rc);
-		async_transaction_destroy(trans);
+	if (trans->request.size > 0) {
+		if ((err = receive_memory_buffer(trans)))
+			goto err;
+	} else {
+		/* The value was valid on the other side, for us, its garbage. */
+		trans->request.buffer.virt = NULL;
 	}
+
+	if ((err = usbhc_iface->transfer(fun, &trans->request,
+	    &transfer_finished, trans)))
+		goto err;
```

Comment:
This is weird. I'm surprised it's even legal to have a label and a variable with the same name.

Reply:
It's pretty common, and perfectly legal - you can even have a `struct err` together with them :) There are several isolated namespaces, labels being one of them.

It might be weird if used in a single instance, but this method of handling error paths is really common, does not clutter the code much and is easily maintainable. Or do you challenge just the name?